### DISCUSSION TITLE
Finding a way to quickly create throwaway test projects for prototyping purposes

### DISCUSSION THREAD

**Discussion by Xrayez** [godot-extended-libraries/godot-ideas#24] [Ideas]
### Related issues

- Add a flag to create a project from the CLI: godotengine/godot#25797.

## Problem
Prototyping in Godot is essential to development process. But currently it takes too many steps to get started writing something:

1. Open project manager...
2. Create project: provide a name, path, renderer...
3. Launch the editor...
4. Create scene, create script, write code, set the scene as main...

We need more convenience!

## Suggestion
The idea is to cut down the number of steps required to start a new (throwaway) project. You still could save the project if you like what you've done. In theory it could also be nice to create some template project from which you'd like to start your development endeavors.

Anything goes, as long as you can "press a button" and edit a new project right away, this would be the best way.

## Solution
See [godot-init](https://gist.github.com/Calinou/cd4c67e2a44377c9fb772500daa467f5) snippets which can be useful for initializing new projects easily. The problem is that you're still required to navigate the filesystem in order to start prototyping or to simply test out some idea.

There's also [GDScript online](https://gdscript-online.github.io/) which can be used for prototyping code in GDScript, but obviously you have no ability to edit an actual project with a scene.

There's also [Godot Engine online](https://godotengine.org/online/godot.tools.html) which could also be used for this purpose, but currently experimental. Takes 10-30 seconds to actually start doing something online.

Also look at third-party project managers like https://gitlab.com/godot-stuff/gs-project-manager.

The solution should likely be specific to operating system to provide the most convenience, like registry keys in Windows, for instance adding [`Open with Godot`](https://gist.github.com/Xrayez/fce615b43d525b73b0bcf2252dde0d78) to the context menu of Windows Explorer.

**Comment by ghost** [comment]
for maximum laziness i would have a "Godot (Dummy Project).exe" file which sets everything up for u (or a .bat file)
or something stupid like that

it would create a project, and it would also open it in the editor, you just need to double click on the exe and ur done

or just a button, which would create a dummy project with a placeholder name, the project would open in the editor instantly

**Comment by Xrayez** [comment]
@RaTcHeT302 on Windows, I've been thinking of automating this using [AHK](https://www.autohotkey.com/) scripting which can be compiled into an `exe` file, one could then assign keyboard shortcuts to instantly run new (or existing) Godot projects that way. I don't know if there are equivalent tools on Linux/Mac for that.

**Comment by ghost** [comment]
> 
> 
> @RaTcHeT302 on Windows, I've been thinking of automating this using [AHK](https://www.autohotkey.com/) scripting which can be compiled into an `exe` file, one could then assign keyboard shortcuts to instantly run new (or existing) Godot projects that way. I don't know if there are equivalent tools on Linux/Mac for that.

it sounds kinda finnicky, personally i would just redesign the project system to allow you to be as lazy as possible by default, i think it would be a nice improvement overall

my mindset is that, you NEED to create a project anyway if you want to make a game, so i don't see why we couldn't automate that step and just create a default project for the user out of the box, when you launch the engine for the first time

otherwise i dunno, maybe just have a big button to create a dummy project, i can't think of anything else right now

you could set a path, where all your prototypes should be saved at, and then they could follow this name scheme

Prototype 1
Prototype 2
Prototype 3

etc...

so you just click the button and now you can go prototype crazy

i would bring this up on the Godot repository but i was silently banned, so i guess the people who maintain the engine are not a fan of the way i talk? i don't sugar coat things and, i want to see the engine improve, but if my existence offends people then, whatever, that's on them (but i'm not going to sit around and agree on everything)

**Comment by ghost** [comment]
i had an idea, i think godot should have a command line argument... of some kind, what it should be, i dunno

![](https://i.imgur.com/8mDxxYU.png)

that way you don't need a whole new .exe, or this finnicky autohotkey thing, that way you can just make a shortcut and you can have it do specific stuff, idk, i haven't really thought this through

a setting would probably make more sense, but i'm not sure what this should be like, i can't really visualize this at the moment

but with a command line, you would only need to create a shortcut, so you wouldn't need to constantly monkey around with the settings themselves

idk
![](https://i.imgur.com/XauF4H4.png)

**Comment by Xrayez** [comment]
Having a button to create prototypes in the project manager would be an improvement in my eyes indeed, especially when I see people use the engine for prototyping purposes rather than actually creating games (I know at least two person who do that, another one writes games in Rust but is perfectly fine using Godot for anything else, including testing out ideas), and according to the development direction of Godot and how certain areas are prioritized (from my experience with talking with Godot core developers):

1 Cater to beginner use cases more.
2. Do not allow to have many knobs to configure how the engine can behave.
3. Some use cases are deemed too specific for most users even if they ***complement*** the engine.

That said, this kind of feature is a must have for Godot *specifically* under those circumstances.

> There's also [Godot Engine online](https://godotengine.org/online/godot.tools.html) which could also be used for this purpose, but currently experimental. Takes 10-30 seconds to actually start doing something online.

Quoting myself, the online tool would certainly benefit from such a feature, having to create a test project manually just to start doing something is not acceptable for such a service. 🙂

The command line option should just be named as `--init` I think, which by the way was worked on some level in godotengine/godot#25925, but got abandoned. Yet one still needs to manually open up the generated project that way... so that would only solve the first part of the equation.

**Comment by Calinou** [comment]
> See godot-init snippets which can be useful for initializing new projects easily. The problem is that you're still required to navigate the filesystem in order to start prototyping or to simply test out some idea.

On Linux, `godot-init $(mktemp -d)` can be used to alleviate this. Beware that your project files will be lost when you restart your PC (they'll be in `/tmp/tmp.<random characters>/`).
